llvm/lib/ExecutionEngine/JITLink/aarch32.cpp

```diff
@@ -248,15 +247,49 @@ Error makeUnexpectedOpcodeError(const LinkGraph &G, const ArmRelocation &R,
               static_cast<uint32_t>(R.Wd), G.getEdgeKindName(Kind)));
 }
 
-template <EdgeKind_aarch32 Kind> bool checkOpcode(const ThumbRelocation &R) {
-  uint16_t Hi = R.Hi & FixupInfo<Kind>::OpcodeMask.Hi;
-  uint16_t Lo = R.Lo & FixupInfo<Kind>::OpcodeMask.Lo;
-  return Hi == FixupInfo<Kind>::Opcode.Hi && Lo == FixupInfo<Kind>::Opcode.Lo;
+static auto &getFixupInfoTableArm() {
+  static constexpr size_t Items = LastArmRelocation - FirstArmRelocation + 1;
+  static std::array<std::unique_ptr<FixupInfoBaseArm>, Items> FixupInfoTableArm;
+  return FixupInfoTableArm;
+}
+
+static auto &getFixupInfoTableThumb() {
+  static constexpr size_t Items = LastThumbRelocation - FirstThumbRelocation + 1;
+  static std::array<std::unique_ptr<FixupInfoBaseThumb>, Items> FixupInfoTableThumb;
+  return FixupInfoTableThumb;
 }
 
-template <EdgeKind_aarch32 Kind> bool checkOpcode(const ArmRelocation &R) {
-  uint32_t Wd = R.Wd & FixupInfo<Kind>::OpcodeMask;
-  return Wd == FixupInfo<Kind>::Opcode;
+template <EdgeKind_aarch32 It, EdgeKind_aarch32 Last, typename TableT>
+void populateFixupInfos(TableT &Table, size_t Idx = 0) {
+  Table[Idx] = std::make_unique<FixupInfo<It>>();
+  if constexpr (It < Last) {
+    constexpr auto Next = EdgeKind_aarch32((Edge::Kind)It + 1);
+    populateFixupInfos<Next, Last>(Table, Idx + 1);
+  }
+}
+
+void populateFixupInfos() {
+  static std::once_flag Flag;
+  std::call_once(Flag, []() {
+    populateFixupInfos<FirstArmRelocation, LastArmRelocation>(getFixupInfoTableArm());
+    populateFixupInfos<FirstThumbRelocation, LastThumbRelocation>(getFixupInfoTableThumb());
+  });
+}
+
+static bool checkOpcode(const ArmRelocation &R, Edge::Kind Kind) {
+  assert(Kind >= FirstArmRelocation && Kind <= LastArmRelocation &&
+         "Edge kind is no Arm relocation");
+  assert(getFixupInfoTableArm().at(Kind - FirstArmRelocation) != nullptr &&
+         "Call populateFixupInfos() before dynamic access");
+  return getFixupInfoTableArm()[Kind - FirstArmRelocation]->checkOpcode(R.Wd);
+}
```

Comment:
The opcode checks are on the hot-path of relocation processing. Doing them dynamically involves a virtual function call. This will add a little overhead. I haven't benchmarked it yet, but my gut feeling is: measurable but acceptable (the amount of code going through this backend won't be gigabytes).

Reply:
Quick check running all `llvm-jitlink` invocations from existing LIT tests and measuring CPU time suggests a 0.67% slowdown. If this is getting too expensive, I think we can still make the opcode checks debug-only.

```
#!/bin/bash
bin/llvm-jitlink -noexec -slab-address 0x76ff0000 -slab-allocate 10Kb -slab-page-size 4096 test/ExecutionEngine/JITLink/AArch32/Output/ELF_static_arm_reloc.s.tmp.o
bin/llvm-jitlink -noexec -slab-address 0x76ff0000 -slab-allocate 10Kb -slab-page-size 4096 -abs external_func=0x76bbe880 test/ExecutionEngine/JITLink/AArch32/Output/ELF_static_thumb_reloc.s.tmp.o
bin/llvm-jitlink -noexec -slab-address 0x76ff0000 -slab-allocate 10Kb -slab-page-size 4096 test/ExecutionEngine/JITLink/AArch32/Output/ELF_static_arm_reloc.s.tmp.o
bin/llvm-jitlink -noexec -slab-address 0x76ff0000 -slab-allocate 10Kb -slab-page-size 4096 -abs external_func=0x76bbe880 test/ExecutionEngine/JITLink/AArch32/Output/ELF_static_thumb_reloc.s.tmp.o
bin/llvm-jitlink -noexec -slab-address 0x76ff0000 -slab-allocate 10Kb -slab-page-size 4096 test/ExecutionEngine/JITLink/AArch32/Output/ELF_static_arm_reloc.s.tmp.o
bin/llvm-jitlink -noexec -slab-address 0x76ff0000 -slab-allocate 10Kb -slab-page-size 4096 -abs external_func=0x76bbe880 test/ExecutionEngine/JITLink/AArch32/Output/ELF_static_thumb_reloc.s.tmp.o
bin/llvm-jitlink -noexec -slab-address 0x76ff0000 -slab-allocate 10Kb -slab-page-size 4096 -abs target=0x76bbe88f test/ExecutionEngine/JITLink/AArch32/Output/ELF_static_data_reloc.s.tmp_armv7.o
bin/llvm-jitlink -noexec -slab-address 0x76ff0000 -slab-allocate 10Kb -slab-page-size 4096 -abs target=0x76bbe88f test/ExecutionEngine/JITLink/AArch32/Output/ELF_static_data_reloc.s.tmp_thumbv7.o
```

Reply:
As the co-maintainer of this backend I think you're in the best position to decide on the performance trade-offs.

I'm going to be pushing for more speed in JITLink in general, especially in the backends I'm responsible for (JITLink is being used in increasingly performance-sensitive areas), but this change would only affect aarch32 so there's no problem there from my perspective.

Reply:
Using explicit function points instead of virtual functions (as you proposed below) gave a significant speedup and the slowdown turned into noise.